src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/JsonFunctionCallUtilities.cs

```diff
@@ -0,0 +1,448 @@
+            return schema;
+        }
+
+        if (serializerOptions is null)
+        {
+            return _trueJsonSchema;
+        }
+
+        return InferParameterJsonSchema(
+            parameterMetadata.ParameterType,
+            parameterMetadata.Name,
+            serializerOptions,
+            description: parameterMetadata.Description,
+            hasDefaultValue: parameterMetadata.HasDefaultValue,
+            defaultValue: parameterMetadata.DefaultValue,
+            inferenceOptions);
+    }
+
+    /// <summary>
+    /// Determines a JSON schema for the provided parameter metadata.
+    /// </summary>
+    /// <param name="type">The type of the parameter.</param>
+    /// <param name="parameterName">The name of the parameter.</param>
+    /// <param name="serializerOptions">The options used to extract the schema from the specified type.</param>
+    /// <param name="description">The description of the parameter.</param>
+    /// <param name="hasDefaultValue">Whether the parameter is optional.</param>
+    /// <param name="defaultValue">The default value of the optional parameter, if applicable.</param>
+    /// <param name="inferenceOptions">The options controlling schema inference.</param>
+    /// <returns>A JSON schema document encoded as a <see cref="JsonElement"/>.</returns>
+    public static JsonElement InferParameterJsonSchema(
+        Type? type,
+        string parameterName,
+        JsonSerializerOptions serializerOptions,
+        string? description = null,
+        bool hasDefaultValue = false,
+        object? defaultValue = null,
+        JsonSchemaInferenceOptions? inferenceOptions = null)
+    {
+        _ = Throw.IfNull(parameterName);
+        _ = Throw.IfNull(serializerOptions);
+        inferenceOptions ??= JsonSchemaInferenceOptions.Default;
+
+        FunctionParameterKey key = (
+            type,
+            parameterName,
+            description,
+            hasDefaultValue,
+            defaultValue,
+            IncludeSchemaUri: false,
+            inferenceOptions.DisallowAdditionalProperties,
+            inferenceOptions.IncludeTypeInEnumSchemas);
+
+        return GetJsonSchemaCached(serializerOptions, key);
+    }
+
+    /// <summary>Infers a JSON schema for the specified type.</summary>
+    /// <param name="type">The type for which to generate the schema.</param>
+    /// <param name="serializerOptions">The options used to extract the schema from the specified type.</param>
+    /// <param name="description">The description of the parameter.</param>
+    /// <param name="hasDefaultValue">Whether the parameter is optional.</param>
+    /// <param name="defaultValue">The default value of the optional parameter, if applicable.</param>
+    /// <param name="inferenceOptions">The options controlling schema inference.</param>
+    /// <returns>A <see cref="JsonElement"/> representing the schema.</returns>
+    public static JsonElement InferJsonSchema(
+        Type? type,
```

Comment:
Why do we allow a null Type?

Reply:
`AIFunctionMetadata` allows it so this needs to allow it as well.

Reply:
> I would expect the handling of a null Type to live in the function that accepts an AIFunctionMetadata.

`AIFunctionMetadata` itself can contain schemas, so any function factory building the metadata needs to use an overload doesn't require it as input.

Reply:
Both public APIs can use the same internal code. It'd just be the public one accepting a Type that would throw for a null Type before delegating the rest of its processing to the internal worker.

What's the scenario where we would encourage someone to call this public method with a null Type?

Reply:
> AIFunctionMetadata itself can contain schemas, so any function factory building the metadata needs to use an overload doesn't require it as input.

What's the scenario where someone would be building up an AIFunctionMetadata with a null Type? I'm genuinely asking, I'm missing it.